kong/api/endpoints.lua

```diff
@@ -222,7 +222,11 @@ local function query_entity(context, self, db, schema, method)
       return dao[context](dao, size, args.offset, opts)
     end
 
-    return dao[method](dao, self.params[schema.name], size, args.offset, opts)
+    local key = self.params[schema.name]
+    if type(key) == "string" then
```

Comment:
I really don't like where in the call stack `unescape_uri` is called - It seems that it happens way too late, and should rather be done at the "web framework" level.  I put it here because for select/insert/update/upsert/delete, the path component is unescaped as well, so it seems to fit into the overall scheme.  I do think, however, that unescaping the path components is the right thing for all handlers, not just for tags.

I think that one alternative would be to unescape the tags in `get_collection_endpoint`, but as that uses the generic `page_collection` and `query_entity` functions, it would have to modify `self.params[schema.name]`, which I'd find to be surprising.  It would allow us to localize the escapinging in one place, but it would also mean that next time someone wants to allow spaces in path components, they'd run into this issue again and maybe not even find the specialized solution for tags.  A more complete solution would probably have to consider multiple path parameters, but at least that would be relatively easy to add.

Reply:
My comment was not about the place where we do the encoding (which, I agree, feels weird, like it should happen immediately when the params are converted into Lua tables, but ok).

It was about *the kind of condition that we use to detect that encoding is needed*. On one hand it is:

```
if type(args.tags) == "table" and args.tags[1] or args.tags
```

But here it's just:

```
if type(key) == "string" then
```
Which is more "generic" than the other one, which is clearly associated to "Tags".

Perhaps we can use this?

```suggestion
    if schema.name == "tags" then
```

Reply:
I have tried replacing this condition on a separate branch and the tests seem happy about it:

https://github.com/Kong/kong/tree/FT-2680-space-in-tags-condition